File: golang/solid/single_responsibility.go
package solid

import (
	"fmt"
	"net/url"
)

// Journal struct with responsibility to manage entries
type Journal struct {
	Entries []string
}

var entryCount = 0

// AddEntry adds entry to the entries for the Journal (single responsibility is ok)
func (j *Journal) AddEntry(entry string) int {
	entryCount++
	j.Entries = append(j.Entries, fmt.Sprintf("%v: %v", entryCount, entry))
	return entryCount
}

// RemoveEntry removes entry from the entries of the Journal (single responsibility is ok)
func (j *Journal) RemoveEntry(index int) {
	entryCount--
	j.Entries = append(j.Entries[:index], j.Entries[index+1:]...)
}

// SaveToFile breaks the single responsibility
func (j *Journal) SaveToFile(fileName string) {}

// LoadFromFile breaks the single responsibility
func (j *Journal) LoadFromFile(fileName string) {}

/*
	We should separate the persistence from the journal struct
	by separating to a separated struct or stand-alone functions
*/

type Persistence struct {
	LineSeparator string
}

func (p *Persistence) SaveToFile(fileName string) {}

func (p *Persistence) LoadFromFile(fileName string) {}

func (p *Persistence) LoadFromWeb(url *url.URL) {}

/*
	Now we moved the persistence to separated component
	and everything is ok now (we could do stand-alone functions)
*/

File: python/solid/single_responsibility.py
class Journal:
    entries = []
    count = 0

    def add_entry(self, text):
        self.entries.append(f'{self.count}: {text}')
        self.count += 1

    def remove_entry(self, index):
        self.entries.pop(index)
        self.count -= 1

    # the persistence methods break the single responsibility
    def save_to_file(self, file_name, line_separator):
        pass

    def load_from_file(self, file_name):
        pass


# should separate the persistence (another class or just functions)
class Persistence:
    def __init__(self, line_separator):
        self.line_separator = line_separator

    def save_to_file(self, file_name):
        pass

    def load_from_file(self, file_name):
        pass
